### DISCUSSION TITLE
HTTPS client is not reading all the data from socket

### DISCUSSION THREAD

**Discussion by d-makarenko** [zephyrproject-rtos/zephyr#63030] [General] [answered]
I am playing with `samples/net/sockets/http_client` on nucleo-f207zg. I am using Zephyr 3.3.

While the plain HTTP client on ipv4 works fine, I have some issues with HTTPS. When downloading a file over a HTTPS/TLSv1.2 connection, the last TLS packet is not processed and pushed into user callback.
After some debugging, I found something that was not clear to me in `zephyr/subsys/net/lib/http/http_client.c:447`

```c
		ret = zsock_poll(fds, nfds, remaining_time);
		if (ret == 0) {
			LOG_DBG("Timeout");
			goto finalize_data;
		} else if (ret < 0) {
			goto error;
		}
		if (fds[0].revents & (ZSOCK_POLLERR | ZSOCK_POLLNVAL)) {
			goto error;
		} else if (fds[0].revents & ZSOCK_POLLHUP) {
			/* Connection closed */
			LOG_DBG("Connection closed");
			goto finalize_data;
		} else if (fds[0].revents & ZSOCK_POLLIN) {
```

There is a check if the socket is in a POLLHUP state, and if it is, the connection is considered closed, and no more data is being read from the socket. According to the POSIX manual there could be data to read:
```
       POLLHUP
              Hang up (only returned in revents; ignored in events).
              Note that when reading from a channel such as a pipe or a
              stream socket, this event merely indicates that the peer
              closed its end of the channel.  Subsequent reads from the
              channel will return 0 (end of file) only after all
              outstanding data in the channel has been consumed.
```

More so, if I comment out this check, I can successfully read the whole file without any issues.

So, does this code work correctly? Do I have some misconfiguration somewhere? Or is it a bug?

**Comment by nordicjm** [comment]
@rlubos

**Comment by rlubos** [comment, marked as answer]
Hi @d-makarenko, it seems you're right, we should still call `recv()` after getting `ZSOCK_POLLHUP`, so it looks like a bug on the HTTP client side. On second thought, I think we may even skip the `ZSOCK_POLLHUP` at all, as it should always be accompanied with `ZSOCK_POLLIN`, and leave it up to `recv()` to detect connection close. Would you be willing to send a PR to fix that?

**Reply by d-makarenko** [reply]
Sure, I agree that there is nothing we can do special about ZSOCK_POLLHUP, so it makes sense to skip it.
I will create a PR.

Thanks for the quick response!

**Comment by patrykkk1** [comment]
Hi guys, I've also encountered the same issue as @d-makarenko. Could you please provide an update on the progress? Has the pull request been created?
Thanks, best regards
